bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/TableImpl.java

```diff
@@ -0,0 +1,127 @@
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.models.v1;
+
+import com.adobe.cq.export.json.ComponentExporter;
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.wcm.core.components.models.Table;
+import com.adobe.cq.wcm.core.components.services.table.ResourceProcessor;
+import com.day.cq.dam.api.Asset;
+import com.day.cq.dam.commons.util.DamUtil;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.models.annotations.Exporter;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.Optional;
+import org.apache.sling.models.annotations.injectorspecific.InjectionStrategy;
+import org.apache.sling.models.annotations.injectorspecific.SlingObject;
+import org.apache.sling.models.annotations.injectorspecific.ValueMapValue;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Objects.nonNull;
+
+@Model(adaptables = SlingHttpServletRequest.class,
+    adapters = {Table.class, ComponentExporter.class},
+    resourceType = TableImpl.RESOURCE_TYPE)
+
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME,
+    extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class TableImpl extends AbstractComponentImpl implements Table {
+
+    public static final String RESOURCE_TYPE = "core/wcm/components/table/v1/table";
+    @ValueMapValue(name = "source", injectionStrategy = InjectionStrategy.OPTIONAL)
+    private String source;
+
+    @ValueMapValue(name = "headerNames", injectionStrategy = InjectionStrategy.OPTIONAL)
+    private String[] headerNames;
+
+    @ValueMapValue(name = "description", injectionStrategy = InjectionStrategy.OPTIONAL)
+    private String description;
+
+    @Inject @Optional
+    private List<ResourceProcessor> resourceProcessors;
+
+    @SlingObject
+    private ResourceResolver resourceResolver;
+
+    private List<String> formattedTableHeaderNames;
+    private List<List<String>> items;
+
+
+    @PostConstruct
+    public void initModel() {
+        formatHeaderNames();
+    }
+
+    private String getSourceResourceMimeType(Resource sourceResource) {
+        if (nonNull(sourceResource)) {
+            if (DamUtil.isAsset(sourceResource)) {
+                Asset asset = DamUtil.resolveToAsset(sourceResource);
```

Comment:
asset can, in theory, be null because `DamUtil.resolveToAsset` uses `adaptTo` which doesn't guarantee a non-null return.

Reply:
@ky940819 DamUtil.isAsset(sourceResource) check is to ensure that the sourceResource is an asset. If it is Asset then it should adaptTo Asset with no issues. It can throw null if it is not adaptable to Asset which is not the case here.

Reply:
`resolveToAsset` also performs an `isAsset` check; by performing your own check you are only preventing the parent resolution if the current resource is a child of an asset. 

Either way, `isAsset` only confirms the resource type is "dam:Asset". It makes no promise that `Resource.adaptTo(Asset.class)` will succeed. 
It is never safe to assume that `Adaptable.adaptTo()` will return a non-null value.

You and I both know that a resource of type "dam:Asset" will --under normal conditions, in most cases -- return an Asset. 
However, the Sling API explicitly states that it is acceptable for `adaptTo()` to return null, so we always have to assume that will happen at some point and check for null.

Reply:
Thanks for the quick responses, I really appreciate it. Certainly adding Null check is the safest way,  I will add it here.
Do you think any scenario, where asset type is `dam:Asset` and fails to adapt to Asset?